PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc

```diff
@@ -0,0 +1,111 @@
+        : candidateToken_(consumes<edm::View<pat::PackedCandidate>>(iConfig.getParameter<edm::InputTag>("candidates"))),
+          trackToken_(consumes<edm::View<reco::Track>>(iConfig.getParameter<edm::InputTag>("trackCollection"))),
+          doLostTracks_(iConfig.getParameter<bool>("doLostTracks")) {
+      if (doLostTracks_) {
+        track2LostTrackToken_ = consumes<edm::Association<pat::PackedCandidateCollection>>(
+            iConfig.getParameter<edm::InputTag>("candidates"));
+      } else {
+        candidate2PFToken_ =
+            consumes<edm::Association<reco::PFCandidateCollection>>(iConfig.getParameter<edm::InputTag>("candidates"));
+      }
+
+      produces<FloatMap>();
+    }
+    ~PackedCandidateTrackChi2Producer() override{};
+
+    void produce(edm::Event&, const edm::EventSetup&) override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    edm::EDGetTokenT<edm::View<pat::PackedCandidate>> candidateToken_;
+    edm::EDGetTokenT<edm::Association<reco::PFCandidateCollection>> candidate2PFToken_;
+    edm::EDGetTokenT<edm::Association<pat::PackedCandidateCollection>> track2LostTrackToken_;
+    edm::EDGetTokenT<edm::View<reco::Track>> trackToken_;
+    bool doLostTracks_;
+  };
+
+}  // namespace pat
+
+void pat::PackedCandidateTrackChi2Producer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<edm::View<pat::PackedCandidate>> candidates;
+  iEvent.getByToken(candidateToken_, candidates);
+
+  edm::Handle<edm::Association<reco::PFCandidateCollection>> candidate2PF;
+  if (!doLostTracks_) {
+    iEvent.getByToken(candidate2PFToken_, candidate2PF);
+  }
+
+  edm::Handle<edm::Association<pat::PackedCandidateCollection>> tracks2LT;
+  edm::Handle<edm::View<reco::Track>> trks;
+  if (doLostTracks_) {
+    iEvent.getByToken(track2LostTrackToken_, tracks2LT);
+    iEvent.getByToken(trackToken_, trks);
+  }
+
+  const auto& nCand = candidates->size();
+  std::vector<float> trkChi2Map(nCand, 0);
+
+  for (size_t i = 0; i < nCand; i++) {
+    const auto& cand = candidates->refAt(i);
+    float nChi2 = 0;
+
+    // ignore neutral candidates or without track
+    if (cand->charge() == 0 || !cand->hasTrackDetails())
+      continue;
+
+    if (doLostTracks_) {  //for Lost tracks we don't have references to PFCands, so we must loop over tracks and check keys...
+      for (size_t j = 0; j < trks->size(); j++) {
+        const auto& trk = trks->refAt(j);
+        const auto& lostTrack = (*tracks2LT)[trk];
+        if (lostTrack.isNonnull() && (cand.id() == lostTrack.id()) && (cand.key() == lostTrack.key())) {
+          nChi2 = trk->normalizedChi2();
+          break;
+        }
+      }
+    } else {  //for the regular PackedPFCands we have direct references...
+      const auto& candTrack = (*candidate2PF)[cand]->trackRef();
+      nChi2 = candTrack->normalizedChi2();
+    }
+
+    trkChi2Map.at(i) = nChi2;
```

Comment:
For quantities stored in MINIAOD, we should aim for the minimal necessary precision, as is done in e.g. https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATMuonSlimmer.cc#L150. You mention that uint8 was not sufficient in https://github.com/cms-sw/cmssw/pull/30566#issuecomment-654991843. What is the minimum number of significant digits that's needed for this variable when stored as a reduced-size float?

Going from the default of 23 to 12 bits for the mantissa gives an average relative error of ~1e-6, so perhaps you can reduce even more.